### ISSUE TITLE
Spark LightGBM support for re-training from the existing model 

### ISSUE THREAD

**Issue by sundeepks** [microsoft/SynapseML#1529]
Do we have a support for checkpointing/saving the model & re-use the already trained model to train against the next day dataset

I have gone over the docs but did not any around the same..any pointers would be of great help 

Although I have come across setInitScoreCol for the same & any example of the usage would be helpful

AB#1829059

**Comment by svotaw** [comment]
Yes, there is a saveNativeModel(filename) and getNativeModel() to return string as a variable.  With this string you can recreate the model that was previously trained.

```
val fitModel = classifier.fit(df)
val modelStr = fitModel.getModel.modelStr.get
// or fitModel.saveNativeModel(filename)
```
To recreate model for use to transform, simply load the modelStr
```
      val resultsFromString = LightGBMClassificationModel
        .loadNativeModelFromString(modelStr)
       // or .loadNativeModelFromFile(filename)
        .transform(df)
```
To recreate the model for further training:
```
val classifier = new LightGBMClassifier()
  .setModelString(modelStr)
  .setInitScoreCol(initScoreColName)
```

The dataframe column for initial scores (initScoreColName above) should contain a single double for 1 class or a Vector of doubles for multiclass (1 per class)

Note that I am right now fixing a bug in using initial scores for multiclass classification, so if you are doing that you should wait.  The PR is out and should be checked in today.

Let me know if you have further questions or encounter any issues

**Comment by sundeepks** [comment]
@svotaw to recreate the model for further training is it required to setInitScoreCol or just setting the model string is good enough ?

`val classifier = new LightGBMClassifier().setModelString(modelStr)`

Currently I am saving the model & reading it back for further training like below do let me know if I am missing anything here as I am observing higher binary log loss just want to make sure in the model re-training with below code snippet it's able to reuse saved model file.

Saving the model file 
`model.saveNativeModel(model_path)`

Reading the saved model file back

```
modelLoader = LightGBMClassificationModel.loadNativeModelFromFile(model_path) 
modelstr = modelLoader.getLightGBMBooster().modelStr().get()

classifierModel = LightGBMClassifier(objective="binary", featuresCol="features", labelCol="label", validationIndicatorCol="validation", earlyStoppingRound=25, isUnbalance=True, useSingleDatasetMode=True, numLeaves=256, numIterations=512, useBarrierExecutionMode=True)
classifierModel.setModelString(modelstr)

model = classifierModel.fit(df)
```

**Comment by sundeepks** [comment]
Can you please help on the above observation ?

**Comment by svotaw** [comment]
apologies, I have been on vacation the last week.

Yes, you must set setInitScoreCol().  This is part of our wrapper, and not part of the saved model in native LightGBM SDK.

**Comment by svotaw** [comment]
what version of SynapseML are you using?

**Comment by sundeepks** [comment]
@svotaw following is the version I am using "spark.jars.packages": "com.microsoft.azure:synapseml_2.12:0.9.5"

**Comment by imatiach-msft** [comment]
This looks good to me.  Actually, if you are setting the model string, I don't even think you have to set the init score.  That's used more when you are using a different model, for example say an SVM model or a DNN model, to make an initial prediction and then you want to correct that with the new LightGBM iterations to train new tree models on top of the original model (a new tree model is trained per iteration).  For more info see:
https://datascience.stackexchange.com/questions/17074/what-is-init-score-in-lightgbm 
Since you are passing the original LightGBM model string, you don't even need to pass the init score really, but I don't see how it would hurt your performance metrics in any way.

**Comment by imatiach-msft** [comment]
@svotaw actually, I wonder if there is some way to change the code to call fit multiple times?  I recall I added functionality to train in batches (through numBatches parameter), which just calls the inner train (or fit equivalent) in a loop, to reduce memory usage but still allow lightgbm to run on more data at the cost of accuracy (or whatever performance metric is optimized for):

https://github.com/microsoft/SynapseML/blob/master/lightgbm/src/main/scala/com/microsoft/azure/synapse/ml/lightgbm/LightGBMBase.scala#L58

I would think we should be able to somehow change the API to just call fit multiple times, but right now, I guess, since we return a LightGBMClassificationModel class and not the original LightGBMClassifier class, we don't have a fit on it technically because of the different API, so the user is forced to do this kind of weird hack to get the booster from the trained model and pass that to an untrained model for training further iterations.  I'm not sure what the best way to change the API might be to make this workflow feel less hacky, but it certainly seems like we could do something like that.

**Comment by imatiach-msft** [comment]
Oh, I have an idea - just brainstorming here, maybe user could be allowed to do something like (in python code here, but equivalent could be added in scala):

```
// for first training iteration:
my_dataset = ... <load dataset here>
// for second iteration:
my_new_dataset = ... <load another dataset here>

untrained_model = LightGBMClassifer(...)
model = untrained_model.fit(my_dataset)
// new API proposal: extract an "untrained model" from the LightGBMClassificationModel that is of type LightGBMClassifier again, but which starts from the previous iterations of the trained model
model.untrained_model.fit(my_new_dataset)
```

User could use something like "untrained_model" property to get back a LightGBMClassifier that they could do further training on.  Hmm.

**Comment by imatiach-msft** [comment]
Maybe "untrained_model" is a bad name here.  Perhaps "init_model" might be better, since it is actually a trained model, it's just it's the right class type so it has a fit method and can be retrained.

Another alternative might be to actually add the fit() method API to the LightGBMClassificationModel.  I don't know of any spark learners with this API in the model as opposed to the untrained model, so we would be going against how spark ML structures its learners here in some sense I think, but maybe this would be more convenient and less hacky than the current workflow.

**Comment by sundeepks** [comment]
@imatiach-msft Thanks for the detailed response on init_scores, would prefer easy way of calling fit as u suggested above, It took quite some time to figure how to get the model string (from booster) & set it back.

**Comment by ppruthi** [comment]
@sundeepks -- if you're satisfied with the comments, can you please close the issue ? If you feel there are improvements to be made, can you please file a feature request ?
